src/glib/mine/signalproc.cpp

```diff
@@ -1241,6 +1241,298 @@ PJsonVal TOnlineHistogram::SaveJson() const {
+	for (; I < TempLast; ++I) {
+		Sum += TempWeight[I];
+		K1 = MergeCentroid(Sum, K1, TempWeight[I], TempMean[I]);
+	}
+
+	// only existing centroids remain
+	for (; J < N_; ++J) {
+		Sum += W[J];
+		K1 = MergeCentroid(Sum, K1, W[J], U[J]);
+	}
+
+	TempLast = 0;
+
+	// swap pointers for working space and merge space
+
+	Weight = MergeWeight;
+	Mean = MergeMean;
+
+	MergeMean = U;
+	MergeWeight = W;
+
+	MergeMean[0] = Weight[0];
+	for (I=1, N_= Last, MergeWeight[0]=0; I<=N_; ++I) {
+		MergeWeight[I] = 0; // zero out merge weights
+		MergeMean[I] = MergeMean[I-1] + Weight[I]; // stash cumulative dist
+	}
+
+	//TempMean = TuUnsort;
+
+	Min = TMath::Mx(Min, Mean[0]);
+	Max = TMath::Mx(Max, Mean[N_]);
+}
+
+TFlt TTDigest::MergeCentroid(TFlt Sum, TFlt& K1, TFlt& Wt, TFlt& Ut) {
+	TFlt K2 = Integrate((double)Nc, Sum/TotalSum);
+	if (K2 - K1 <= 1.0 || MergeWeight[Last] == 0.0) {
+		// merge into existing centroid if centroid index difference (k2-k1)
+		// is within 1 or if current centroid is empty
+		MergeWeight[Last] += Wt;
+		MergeMean[Last] += (Ut - MergeMean[Last]) * Wt / MergeWeight[Last];
+	} else {
+		// otherwise create a new centroid
+		Last = ++Last;
+		MergeMean[Last] = Ut;
+		MergeWeight[Last] = Wt;
+		K1 = Integrate((double)Nc, (Sum - Wt)/TotalSum);
+	}
+	return K1;
+};
+
+TFlt TTDigest::Integrate(TFlt Nc, TFlt Q_) const {
+	// First, scale and bias the quantile domain to [-1, 1]
+	// Next, bias and scale the arcsin range to [0, 1]
+	// This gives us a [0,1] interpolant following the arcsin shape
+	// Finally, multiply by centroid count for centroid scale value
+	return Nc * (asin(2 * Q_ - 1) + TMath::Pi / 2) / TMath::Pi;
+}
+
+TInt TTDigest::Bisect(const TFltV& A, const TFlt& X, TInt Low, TInt& Hi) const {
+	while (Low < Hi) {
+		TInt Mid = (Low + Hi) >> 1;
+		if (A[Mid] < X) {
+			Low = Mid + 1;
+		}
+		else {
+			Hi = Mid;
+		}
+	}
+	return Low;
+}
+
+TFlt TTDigest::Interp(const TFlt& X, const TFlt& X0, const TFlt& X1) const {
+	TFlt Denom = X1 - X0;
+	return Denom > EPSILON ? (X - X0) / Denom : 0.5;
+}
+
+TFlt TTDigest::Boundary(const TInt& I, const TInt& J, const TFltV& U, const TFltV& W) const {
+	return U[I] + (U[J] - U[I]) * W[I] / (W[I] + W[J]);
+}
+
+TInt TTDigest::NumTemp(const TInt& N) const {
+	TInt Lo = 1, Hi = N, Mid;
+	while (Lo < Hi) {
+		Mid = (Lo + Hi) >> 1;
+		if (N > Mid * TMath::Log(Mid) / TMath::Log(2)) {
+			Lo = Mid + 1;
+		}
+		else {
+			Hi = Mid;
+		}
+  }
+  return Lo;
+}
+
+void TTDigest::Print() const {}
+
+void TTDigest::SaveState(TSOut& SOut) const {
+	Nc.Save(SOut);
+	Size.Save(SOut);
+	Last.Save(SOut);
+	TotalSum.Save(SOut);
+	Weight.Save(SOut);
+	Mean.Save(SOut);
+	Min.Save(SOut);
+	Max.Save(SOut);
+	MergeWeight.Save(SOut);
+	MergeMean.Save(SOut);
+	Tempsize.Save(SOut);
+	UnmergedSum.Save(SOut);
+	TempLast.Save(SOut);
+	TempWeight.Save(SOut);
+	TempMean.Save(SOut);
+	DEFAULT_CENTROIDS.Save(SOut);
+	Quantiles.Save(SOut);
+}
+
+void TTDigest::LoadState(TSIn& SIn) {
+	Nc.Load(SIn);
```

Comment:
I don't think all these variables correspond to state. For example, the datalib-sketch only saves: number of centroids, min, max, centers and weights.